gson/src/main/java/com/google/gson/ReflectionAccessFilter.java

```diff
@@ -0,0 +1,187 @@
+ * convenience filters for this task, such as {@link #BLOCK_INACCESSIBLE_JAVA}.
+ *
+ * <p>A filter can also be useful to prevent mixing model classes of a
+ * project with other non-model classes; the filter could
+ * {@linkplain FilterResult#BLOCK_ALL block all reflective access} to
+ * non-model classes.
+ *
+ * <p>A reflection access filter is similar to an {@link ExclusionStrategy}
+ * with the major difference that a filter will cause an exception to be
+ * thrown when access is disallowed while an exclusion strategy just skips
+ * fields and classes.
+ *
+ * @see GsonBuilder#addReflectionAccessFilter(ReflectionAccessFilter)
+ */
+public interface ReflectionAccessFilter {
+  /**
+   * Result of a filter check.
+   */
+  enum FilterResult {
+    /**
+     * Reflection access for the class is allowed.
+     *
+     * <p>Note that this does not affect the Java access checks in any way,
+     * it only permits Gson to try using reflection for a class. The Java
+     * runtime might still deny such access.
+     */
+    ALLOW,
+    /**
+     * The filter is indecisive whether reflection access should be allowed.
+     * The next registered filter will be consulted to get the result. If
+     * there is no next filter, this result acts like {@link #ALLOW}.
+     */
+    INDECISIVE,
+    /**
+     * Blocks reflection access if a member of the class is not accessible
+     * by default and would have to be made accessible. This if unaffected
+     * by any {@code java} command line arguments being used to make packages
+     * accessible.
+     *
+     * <p>Note that this <b>only works for Java 9 and higher</b>, for older
+     * Java versions its functionality will be limited and it might behave like
+     * {@link #ALLOW}. Access checks are only performed as defined by the Java
+     * Language Platform System, restrictions imposed by a {@link SecurityManager}
+     * are not considered.
+     *
+     * <p>This result type is mainly intended to help enforcing the access checks of
+     * the Java Platform Module System. It allows detecting illegal access, even if
+     * the used Java version would only log a warning, or is configured to open
+     * packages for reflection using command line arguments.
+     *
+     * @see AccessibleObject#setAccessible(boolean)
+     * @see AccessibleObject#canAccess(Object)
+     */
+    BLOCK_INACCESSIBLE,
+    /**
+     * Blocks all reflection access for the class. Other means for serializing
+     * and deserializing the class, such as a {@link TypeAdapter}, have to
+     * be used.
+     */
+    BLOCK_ALL
+  }
+
+  /**
+   * Blocks all reflection access to members of standard Java classes which are
+   * not accessible by default. However, reflection access is still allowed for
+   * classes for which all fields are accessible and which have an accessible
+   * no-args constructor (or for which an {@link InstanceCreator} has been registered).
```

Comment:
I think it would be good to spell out here exactly what a "standard Java class" is. Maybe just, at the end of the paragraph:

"A standard Java class is any class in a {@code java.*} or {@code javax.*} package."